### ISSUE TITLE
CNAME resolution of AWS Aurora database cluster address

### ISSUE THREAD

**Issue by stevenscg** [hashicorp/consul#3060]
### `consul version` for both Client and Server
Client: `v0.8.2` and `v0.8.3`
Server: `v0.8.2` and `v0.8.3`

### Operating system and Environment details
CentOS7 x86_64

### Description of the Issue (and unexpected/desired result)

Originally posted in the [mailing list](https://groups.google.com/d/msg/consul-tool/x-eceQxHP-U/REiH8IbOBAAJ).

The Aurora database service provides an address for each specific database instance and a _cluster address_ (a CNAME pointing to the current database leader/writer).

Resolution of this cluster address through Consul stops with a CNAME instead of an A record.

This is on an otherwise working Consul cluster using dnsmasq and the AWS VPC DNS recursor.

### Reproduction steps

A "db" service in Consul configured with the **instance** address/endpoint:
```
; <<>> DiG 9.8.2rc1-RedHat-9.8.2-0.37.rc1.43.amzn1 <<>> db.service.consul
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 487
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;db.service.consul.	IN	A

;; ANSWER SECTION:
db.service.consul. 5	IN	CNAME	db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com.
db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com. 5 IN	CNAME ec2-XXX-XXX-XXX-XXX.us-west-2.compute.amazonaws.com.
ec2-XXX-XXX-XXX-XXX.us-west-2.compute.amazonaws.com. 20 IN A 10.XXX.XXX.XXX
```

A "db" service in Consul configured with the **cluster** address/endpoint:
```
; <<>> DiG 9.8.2rc1-RedHat-9.8.2-0.37.rc1.43.amzn1 <<>> db.service.consul
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 48735
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;db.service.consul.	IN	A

;; ANSWER SECTION:
db.service.consul. 5	IN	CNAME	db-cluster.cluster-XXXXXXXXXXXX.us-west-2.rds.amazonaws.com.
db-cluster.cluster-XXXXXXXXXXXX.us-west-2.rds.amazonaws.com. 5 IN	CNAME db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com.
db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com. 5 IN	CNAME ec2-XXX-XXX-XXX-XXX.us-west-2.compute.amazonaws.com.
```

### Expected result:
Clients connecting via the 
### Actual result:
Clients connecting via the consul dns using this second configuration fail due to the missing A record.

**Comment by slackpad** [comment]
Hi @stevenscg this looks like the recursor is stopping early for that second case.

**Comment by stevenscg** [comment]
@slackpad Yeah, I'd think so too. Can you think of anything in consul or dnsmasq that would be truncating?

Also, a dig on `db-cluster.cluster-XXXXXXXXXXXX.us-west-2.rds.amazonaws.com` from the same host resolves all the way to the IP. For now at least, I have to assume that the same recursor is being used in both cases.

**Comment by slackpad** [comment]
We'd have to dig into Consul's DNS client behavior when it's doing the recursor. That would be separate from what the host is doing via `dig` directly.

**Comment by stevenscg** [comment]
I am about to move some more services to Aurora. I'm curious if anybody else is encountering this or if recent consul versions have solved it?

**Comment by stevenscg** [comment]
I tried this again with Consul 1.0.2. The IP is now returned where it was not before:
```
; <<>> DiG 9.9.4-RedHat-9.9.4-51.el7_4.1 <<>> db-test.service.consul SRV
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 48465
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;db-test.service.consul.		IN	SRV

;; ANSWER SECTION:
db-test.service.consul.	1	IN	SRV	1 1 3306 db-test.node.us1.consul.

;; ADDITIONAL SECTION:
db-test.node.us1.consul. 1	IN	CNAME	db-cluster.cluster-XXXXXXXXXXXX.us-west-2.rds.amazonaws.com.
db-cluster.cluster-XXXXXXXXXXXX.us-west-2.rds.amazonaws.com. 2 IN	CNAME db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com.
db.XXXXXXXXXXXX.us-west-2.rds.amazonaws.com. 2 IN	CNAME ec2-XXX-XXX-XXX-XXX.us-west-2.compute.amazonaws.com.
ec2-XXX-XXX-XXX-XXX.us-west-2.compute.amazonaws.com. 17 IN A 10.XXX.XXX.XXX
```

However, mysql clients are still unable to connect:
```
mysql -h db-test.service.consul -uadmin -p
Enter password: 
ERROR 2005 (HY000): Unknown MySQL server host 'db-test.service.consul' (0)
```

There are 3 CNAMEs before the A record vs just 2 in the working scenario.

Both scenarios have the 1 A record in the ADDITIONAL SECTION.

Could this be hitting `maxRecurseRecords` (==5) in https://github.com/hashicorp/consul/blob/master/agent/dns.go#L639??

It seems shy of the 5, but there might be something else I'm not seeing right now.

**Comment by merlineus** [comment]
Got the same problem. But versions 1.0.2 and 1.0.5 didn't help.

**Comment by stevenscg** [comment]
Thanks for the feedback @merlineus. I've been surprised that more Aurora users haven't found and commented on this issue.

**Comment by merlineus** [comment]
We solved the problem. It is related to UDP response fragmentation.

Set in config:
    "dns_config" : {
      "enable_truncate" : true
    }

**Comment by stevenscg** [comment]
@merlineus I'll give that a try. Are you aware of any downsides?

**Comment by merlineus** [comment]
From docs:

> enable_truncate - If set to true, a UDP DNS query that would return more than 3 records, or more than would fit into a valid UDP response, will set the truncated flag, indicating to clients that they should re-query using TCP to get the full set of records.

Also good to know: https://www.networkworld.com/article/2231682/cisco-subnet/cisco-subnet-allow-both-tcp-and-udp-port-53-to-your-dns-servers.html

**Comment by stevenscg** [comment]
I just rolled out Consul 1.0.6 in a test environment with `enable_truncate: true` and can confirm that the Aurora cluster reader and writer endpoints can now be configured as services.

@slackpad We can close this _unless_ you still think that there is more "api cleanup" or some other changes that the team wants to look at first.
